ledger/streaming/client/src/block_validator.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Stream that validates block components
+
+use futures::{future, Stream, StreamExt};
+use mc_common::{
+    logger::{log, Logger},
+    HashSet,
+};
+use mc_ledger_db::Ledger;
+use mc_ledger_streaming_api::{
+    BlockStream, BlockStreamComponents, Error as StreamError, Result as StreamResult,
+};
+use mc_transaction_core::{compute_block_id, ring_signature::KeyImage, Block};
+
+/// Create stream factory for validating individual blocks within a stream.
+/// Valid blocks will passed on, blocks that don't pass will pass an error.
+pub struct BlockValidator<US: BlockStream + 'static, L: Ledger + 'static> {
+    upstream: US,
+    start_block: Block,
+    ledger: Option<L>,
```

Comment:
Why is the ledger optional here? IIUC this validation doesn't work without a Ledger

Reply:
It's because. If there is no ledger (i.e. if it's a pass through component) and someone wants to validate the blocks coming from a peer, they can. The ledger step is required if you're somehow writing into a ledger.